modules/distributor/forwarder/otlpgrpc/forwarder.go

```diff
@@ -0,0 +1,164 @@
+	"go.opentelemetry.io/collector/pdata/ptrace"
+	"go.opentelemetry.io/collector/pdata/ptrace/ptraceotlp"
+	"go.uber.org/multierr"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/credentials/insecure"
+
+	"github.com/grafana/tempo/pkg/tempopb"
+)
+
+type Forwarder struct {
+	cfg         Config
+	logger      log.Logger
+	connections map[string]*grpc.ClientConn
+	clients     map[string]ptraceotlp.Client
+	initialized bool
+	mu          *sync.RWMutex
+}
+
+func NewForwarder(cfg Config, logger log.Logger) (*Forwarder, error) {
+	if err := cfg.Validate(); err != nil {
+		return nil, fmt.Errorf("failed to validate config: %w", err)
+	}
+
+	return &Forwarder{
+		cfg:         cfg,
+		logger:      logger,
+		connections: make(map[string]*grpc.ClientConn),
+		clients:     make(map[string]ptraceotlp.Client),
+		initialized: false,
+		mu:          &sync.RWMutex{},
+	}, nil
+}
+
+// Dial creates client connections and clients based on config.
+// Dial is expected to be called only once.
+func (f *Forwarder) Dial(ctx context.Context, opts ...grpc.DialOption) error {
+	f.mu.Lock()
+	defer f.mu.Unlock()
+
+	if f.initialized {
+		return errors.New("already initialized")
+	}
+
+	connections := make(map[string]*grpc.ClientConn)
+	clients := make(map[string]ptraceotlp.Client, len(f.cfg.Endpoints))
+	for _, endpoint := range f.cfg.Endpoints {
+		client, conn, err := f.newTraceOTLPGRPCClientAndConn(ctx, endpoint, f.cfg.TLS, opts...)
+		if err != nil {
+			return fmt.Errorf("failed to create new trace otlp grpc client: %w", err)
+		}
+
+		connections[endpoint] = conn
+		clients[endpoint] = client
+	}
+
+	f.connections = connections
+	f.clients = clients
+	f.initialized = true
+
+	return nil
+}
+
+func (f *Forwarder) ForwardBatches(ctx context.Context, trace tempopb.Trace) error {
+	// tempopb.Trace is wire-compatible with OTLP, so we convert to bytes, and back.
+	m, err := trace.Marshal()
+	if err != nil {
+		return fmt.Errorf("failed to marshal otlp: %v", err)
+	}
+
+	traces, err := ptrace.NewProtoUnmarshaler().UnmarshalTraces(m)
+	if err != nil {
+		return fmt.Errorf("failed to unmarshal traces: %w", err)
+	}
+	req := ptraceotlp.NewRequestFromTraces(traces)
+
+	var errs []error
+	f.mu.RLock()
+	for _, client := range f.clients {
+		if err := f.exportTracesUsingClient(ctx, client, req); err != nil {
+			errs = append(errs, err)
+		}
+	}
+	f.mu.RUnlock()
+
+	return multierr.Combine(errs...)
+}
+
+func (f *Forwarder) Shutdown(_ context.Context) error {
+	f.mu.Lock()
+	defer f.mu.Unlock()
+
+	var errs []error
+	for endpoint, conn := range f.connections {
+		if err := conn.Close(); err != nil {
+			errs = append(errs, fmt.Errorf("failed to close grpc connection for endpoint=%s: %w", endpoint, err))
+		}
+
+		delete(f.connections, endpoint)
+	}
+
+	return multierr.Combine(errs...)
+}
+
+func (f *Forwarder) exportTracesUsingClient(ctx context.Context, client ptraceotlp.Client, req ptraceotlp.Request) error {
```

Comment:
Now that sending traces is sync we could remove this fn for simplicity

Reply:
Done in [2de3a5c](https://github.com/grafana/tempo/pull/1775/commits/2de3a5c157ebb23eec7d1f21af567751ace2b6ef).